### PR TITLE
(enh) deal with loss/error on X display

### LINKED ISSUE #2: Support conditionality based on WM_NAME
Support keymap and modmaps that are conditional on window name, not just class.

Related: https://github.com/tridactyl/tridactyl/issues/4256

I dislike how this is speced in https://github.com/mooz/xkeysnail/pull/129 :

```
define_keymap(lambda wm_class, device_name, wm_name: "Google Docs" in wm_name, {
     K("M-x"): K("M-slash")
})
```

This "just add another param to lambda" isn't great to read and is hard to deal with on the other side due to the varying number of arguments. (not to mention remembering the ordering, etc).

---

So I definitely think we should support this, it's just a question of what syntax to use... I'm perhaps imagining a wrapper function, but open to other suggestions.

```py
conditions(lambda wm_name: "Google Docs" in wm_name, 
   define_keymap("Google Docs", {
       K("M-x"): K("M-slash")
   })
)
```

Or even better a single context dict:

```py
conditional(lambda ctx: "Google Docs" in ctx.wm_name , 
   keymap("Google Docs", {
       K("M-x"): K("M-slash")
   })
)
```

So instead of `modmap` and `conditional_modmap` there would only be a single `modmap` API that you would either wrap in a conditional or not...

### LINKED ISSUE #40: SAFE MODE: Should still proxy keys (no mapping) even if X display cannot be connected
I wonder though if there is an issue with this turning on/off in the MIDDLE of keystrokes... if in the long run we might need something a bit more structured - or perhaps this doesn't happen often enough that it's worth spending a ton of time on?

### PR DESCRIPTION
**Pull request by joshgoebel**
Resolves #2.  Resolves #40.

- Catches errors raised if the X display cannot be connected.
- If the X display cannot be connected the keymapper falls into a "safe mode" where it stops all mapping and merely makes sure input and output continue to function properly.

### CODE REVIEW

**Comment by joshgoebel** [comment]
@RedBearAK   Thanks for all your help - you wouldn't want to test this PR with some of your tough logging in/out/whatever tests?  Hopefully after this lands I can release 0.6 and then take a break for a bit. :)

**Comment by RedBearAK** [comment]
> @RedBearAK Thanks for all your help - you wouldn't want to test this PR with some of your tough logging in/out/whatever tests? Hopefully after this lands I can release 0.6 and then take a break for a bit. :)

🤔 
I was thinking about that, but not sure if I'm actually running `keyszer` the way you are intending, since I'm still just running it in the `venv` in a terminal tab and doing `git pull` all the time to keep up with new changes. I presume that means I'm still running it "as my user" at this point. 

If I pip install `keyszer` a this point, won't it still be running as my user, unless I pip install with `sudo` to install system-wide? And even then won't I have to find a way to launch it as the `keymapper` user upon login, to see what it will do when I attempt to log out or switch users? 

In short I'm not clear on exactly what `keyszer` is now designed to do automatically with regards to running as a special user and dealing with the X errors. 

Or is most of this irrelevant to just trying to test the X error handling mechanism?

**Comment by joshgoebel** [comment]
> Or is most of this irrelevant to just trying to test the X error handling mechanism?

I think this... whichever way you were running it before and having all the problems - run in that way again.  What we're looking for is no more problems.

**Comment by RedBearAK** [comment]
> I think this... whichever way you were running it before and having all the problems - run in that way again. 

Well, I had issues whether I was running as root or as user, so... 

Alright, well, I took a leap of faith and tried "Switch User" from the user menu. I'm in Ubuntu 22.04, GNOME 42.1. In the past the "Switch User" was broken from this user for reasons unknown, but it looks like the upgrade to 22.04 LTS un-broke it. 

Other than the fact that GNOME acted like I had just asked it to suddenly cure cancer and end world hunger when I hit "Switch User" (it took a WHILE and did weird things to the screen), things have appeared to go pretty well. 

At first it was just asking me to log back in to my user, but I clicked on the icon in the corner that lets you select the session type, and suddenly I was back at the full multi-user login screen (where I was supposed to be in the first place). Even before that, I was able to type just fine into the password field. So that was a marked improvement. 

At the full login screen, I was able to select a user with arrows and Enter, then enter the password without difficulty. Once I logged into the other user, I was able to open a text editor and type away with no issues. 

I opened a terminal and saw `keyszer` still running away in the background (as the user I had switched away from), but no longer apparently trying to take over the keyboard while the other user was in X. 

And then I logged out of that user and logged back into this one. Did NOT need to restart anything in order to have my shortcuts active once again. It's as if I never left. 

What I didn't try is installing and running `keyszer` separately while in the other user's desktop, either as user or as root. But at this point I kind of suspect that would work just fine. Each process only actively remapping when it knows it's connected to the X session being shown on screen (or whatever is actually happening). 

I think you'd be OK to allow yourself one cookie for the day, for being a Good Coder[TM] and accomplishing a minor miracle or two. Enjoy it while you can, tomorrow will probably be a disaster. 🤘🏽 😈 🤘🏽 

> What we're looking for is no more problems.

Remember to toss some salt over your shoulder when you say things like that.

**Comment by RedBearAK** [comment]
Oh, I forgot to checkout the `xorg_errors` branch before doing all that. So the necessary bits must already be in main. If there's some specific aspect of the unmerged commits left in this branch that you wanted tested, I'll have to do it again.

**Comment by joshgoebel** [comment]
Nope. I have not merged yet.

**Comment by joshgoebel** [comment]
> What I didn't try is installing and running keyszer separately while in the other user's desktop, either as user or as root. But at this point I kind of suspect that would work just fine. Each process only actively remapping when it knows it's connected to the X session being shown on screen (or whatever is actually happening).

There is no magic here... Running two is not going to work.  The remapper would be stepping all over itself.  If the X display is the same then the new user that logged in would have all their keys mapped also (provided they gave X permissions to the keymapper user).

The keymapper works when it can access the X display it was told about at startup (via the ENV)... when it can't it goes dormant and just passes keys straight thru.
